File: _python_backup/services/parsing_service.py
"""
Parsing service for VNG files
"""

from typing import Dict, Any
from domain.models import ParsedFile, MetricValue
from domain.exceptions import ParsingError
from modules.parser import parse_vng_text as _parse_vng_text


class ParsingService:
    """Service for parsing VNG text files"""
    
    @staticmethod
    def parse_file(file_name: str, file_content: str, size_bytes: int = 0) -> ParsedFile:
        """
        Parse a VNG file and return a ParsedFile domain model
        
        Args:
            file_name: Name of the file
            file_content: Raw text content of the file
            size_bytes: Size of the file in bytes
            
        Returns:
            ParsedFile domain model
            
        Raises:
            ParsingError: If parsing fails
        """
        try:
            raw_data = _parse_vng_text(file_content)
            
            # Convert to domain models
            parsed_data: Dict[str, Dict[str, MetricValue]] = {}
            for category, metrics in raw_data.items():
                parsed_data[category] = {
                    metric: MetricValue(
                        value=data['value'],
                        is_flagged=data['is_flagged']
                    )
                    for metric, data in metrics.items()
                }
            
            return ParsedFile(
                name=file_name,
                data=parsed_data,
                size_bytes=size_bytes
            )
        except Exception as e:
            raise ParsingError(f"Failed to parse file {file_name}: {str(e)}") from e
    
    @staticmethod
    def parse_to_dict(file_content: str) -> Dict[str, Dict[str, Dict[str, Any]]]:
        """
        Parse VNG file content to dictionary format (legacy compatibility)
        
        Args:
            file_content: Raw text content of the file
            
        Returns:
            Dictionary with parsed data
        """
        return _parse_vng_text(file_content)

File: js/src/services/parsingService.js
/**
 * Parsing service for VNG files - JavaScript version
 */

const { ParsedFile, MetricValue } = typeof module !== 'undefined' && module.exports
    ? require('../domain/models')
    : window.VNGDomain;

const { ParsingError } = typeof module !== 'undefined' && module.exports
    ? require('../domain/exceptions')
    : window.VNGExceptions;

const { parseVngText } = typeof module !== 'undefined' && module.exports
    ? require('../modules/parser')
    : window.VNGParser;

/**
 * Service for parsing VNG text files
 */
class ParsingService {
    /**
     * Parse a VNG file and return a ParsedFile domain model
     * @param {string} fileName - Name of the file
     * @param {string} fileContent - Raw text content of the file
     * @param {number} sizeBytes - Size of the file in bytes
     * @returns {ParsedFile} ParsedFile domain model
     */
    static parseFile(fileName, fileContent, sizeBytes = 0) {
        try {
            const rawData = parseVngText(fileContent);

            // Convert to domain models
            const parsedData = {};

            for (const [category, metrics] of Object.entries(rawData)) {
                parsedData[category] = {};

                for (const [metric, data] of Object.entries(metrics)) {
                    parsedData[category][metric] = new MetricValue(
                        data.value,
                        data.isFlagged
                    );
                }
            }

            return new ParsedFile(fileName, parsedData, sizeBytes);
        } catch (error) {
            throw new ParsingError(`Failed to parse file ${fileName}: ${error.message}`);
        }
    }

    /**
     * Parse VNG file content to dictionary format (legacy compatibility)
     * @param {string} fileContent - Raw text content of the file
     * @returns {Object} Dictionary with parsed data
     */
    static parseToDict(fileContent) {
        return parseVngText(fileContent);
    }
}

// Export service
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { ParsingService };
} else if (typeof window !== 'undefined') {
    window.VNGServices = window.VNGServices || {};
    window.VNGServices.ParsingService = ParsingService;
}
